### ISSUE TITLE
Multiple localizations running in the same database

### ISSUE THREAD

**Issue by mileo** [odoo/odoo#17778]
I would like to discuss whether Odoo SA has any planning related to the matter.

@dreispt @aricaldeira @rvalyi

**Comment by Yenthe666** [comment]
Hi @mileo,

Could you please give more feedback about what exactly you want or want to know more about? There is very little information in your topic..
What exactly do you mean with multiple localizations? You can easily use multiple langs etc so not sure what exactly you want.

Regards,
Yenthe

**Comment by aricaldeira** [comment]
Hello, @Yenthe666 ,

Today, each country creates its own localisation, that may conflict, and often times do conflict, with the localisation of other countries;

Views are changed, methods modified, required fields are inserted, and all that, for all practical terms, makes impossible for an organisation that needs to work in more than one country use the same Odoo instance to manage both their branches, say, in France and in Brazil, for example.

In Brazil we do have such cases, so we started to scratch our own itch: refactoring our own localisation, and some experimental modules such as:
https://github.com/aricaldeira/odoo-multi-localisation

It creates the concrete concept of a localisation in the system, allows the user to explicitly activate one localisation or another, both on the company model and on the user model, and filters views accordingly, that means, views (inherited or otherwise) meant to be used with one localisation wont interfere with views meant to be used with some another localisation;

The functional part is a lot more trouble than that, and, for now, we've been using a "is_brazilian" boolean field on models that need to behave differently for us, and changing methods' behaviour accordingly, and it seems to be working with no noticeable problems;

@mileo wants to start the discussion with others, to see if they do experience the same issues, whether they have used similar solutions as ours, and so on.

Hopes this makes things more clear now ;)

**Comment by Yenthe666** [comment]
Hi @aricaldeira,

Thank you for adding more details. However, from my point of view, I still don't see any concrete problem or issue. You can have multiple languages, you can change about anything.. Could you please ellaborate one functional case where you have a problem or cannot get something done?

I'm always open for discussion and possible improvements so I'd love to continue this.
Cc @mart-e who is the expert for anything translation / language related.

Regards,
Yenthe

**Comment by mileo** [comment]
Hi, @Yenthe666 

Sorry to have been so brief in creating the issue, but I have been discussing this with @renatonlima and @aricaldeira since December 2016 and to formalize the discussion I created a topic without explaining anything. But let's go:

Some localizations need to make changes to the python code that make major interventions, for example:
https://github.com/OCA/l10n-brazil/blob/10.0/l10n_br_account_product/models/account_invoice.py#L26
In this example the interventions are performed in the calculation of total invoices...

As Odoo has become popular and some multinational customers contacted us in the last year looking for a fully integrated system. We began to wonder how to have complex locations (which need to make deep interventions) installed in a multi-company environment.

Suggestions are welcome!

**Comment by aricaldeira** [comment]
For instance, in Brazil, there are several required fields we must put in partner views; those fields make no sense for the French localisation, so, those required fields break the French localisation; the reverse is possibly true also: how, in the same database, in a multicompany setup, can we make sure that conflicting views, requirements, python code running on server, all that, works as if there were 2 databases, 2 separated instances with mono-company setup?

Then again, I can choose to use Odoo in Portuguese, say, and may still want it to *behave* as a French localised system, say, if the user is Brazilian and the company is in France; or the other way around, use the system in French, but the *behavior* be as if the company where in Brazil;

Imagine you have a manager in France, that is responsible for elaborating some reports about their Brazilian branch for a board meeting; he could log into the system, choose his company to be the Brazilian branch, and all views and methods would respond to that change, activating the localisation of that specific company, and generate accounting reports respecting Brazilian law; or, instead, choose the Brazilian branch, but specify the French localisation, and get those accounting reports respecting French law and configurations;

I don't know if there are countries in Europe that would benefit of such a configuration, but I'm guessing there are;

**Comment by nhomar** [comment]
On Wed, Jun 28, 2017 at 3:45 PM, Aristides Caldeira <
notifications@github.com> wrote:

> Imagine you have a manager in France, that is responsible for elaborating
> some reports about their Brazilian branch for a board meeting; he could log
> into the system, choose his company to be the Brazilian branch, and all
> views and methods would respond to that change, activating the localisation
> of that specific company, and generate accounting reports respecting
> Brazilian law; or, instead, choose the Brazilian branch, but specify the
> French localisation, and get those accounting reports respecting French law
> and configurations;
>
>
What we are doing with l10n_mx_** inside odoo is have methods that have an
explicit:


def l10n_mx_something(self)
    if not company.country is Mexico:
       return
   do your magic....

It has become pretty usefull all localizations should do that, all our 3
localizations are being rewritten using such approach, (l10n_pe, _mx and
_ve) and we use a multicompany enviroment (wip the fiscal part to be
unified).


There are edge cases where this approach does not work obviously, then when
such cases comes in we take another aproaches.

1.- Custom ACL which shows one screen to MX people and another to VE people
with different required fields.
2.- Server actions (a lot of them) doing things non-blocking.
3.- Porst mortem information on mail.message which allows the user continue
the original odoo process but being informed that legally there is some
action to take.

It is a complete change of mind set.

Reagrds.

Nhomar Hernandez
CEO Vauxoo.
Site: http://vauxoo.com
Twitter: @nhomar
Blog: http://nhomar.com
Github User: https://github.com/nhomar
Odoo Gold Partner
Skype: nhomar00 (Envia mail previo no lo superviso siempre).
HangOut: nhomar@vauxoo.com
Móvil Venezuela:
+58 4144110269 <0414-4110269>
Móvil México:
+52 1 4773933942 <+52%201%20477%20393%203942>

**Comment by Yenthe666** [comment]
@mileo and @aricaldeira alright, thank you for the valuable information! Now I see what you guys need and would like.
I honestly think that @mart-e should answer here - he is the most familiar with the languages and can decide about this too. I presume that these changes are to big and advanced to make it into a next Odoo version but who knows.

**Comment by mart-e** [comment]
Hello,

I agree with @nhomar, we usually try to avoid adding blocking code in the localizations like making fields required (or adding new fields only when necessary). The methods you give are good ways to reach your point without blocking other localizations in multicompany.

**Comment by rvalyi** [comment]
Guys I lost the prototype I made some 4 years ago, but basically what I did for the Akretion ERP (with both Brazilian and French localization by the time) was:

-  use one Odoo process in France
-  use another Odoo process for Brazil

both instances were connected to the same database but with a different addons-path per instance. Like the Brazilian localization was loaded only for the Odoo instance with behind the Brazilian domain.

I did a 50 lines patch in Odoo so that during the module initialization module set as installed in the database but not found in the addons-path would not block the ERP start anymore. Instead these modules were recorded in a list in the Odoo registry. Then when serving views and actions (mostly), if some of the database entry had an xml_id showing they belong to one of these deactivated modules, Odoo would skipped these records just like if they didn't exist. Sadly I once trashed this github repo with this branch...

I think the only thing left was to put some default in required fields (in the database) or have a module making these fields only mandatory via the Python layer.

Until Odoo SA really starts addressing this problem at the framework level for complex localizations I think this can be a good workaround in the rare cases where two localizations on the same Odoo database would be a viable option (generally it is not because there is too much coupling even simply in the ERP implementation process, migration etc..).
